### PR TITLE
HBASE-22422 Retain an ByteBuff with refCnt=0 when getBlock from LRUCache

### PR DESCRIPTION
**Pull request by openinx**

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java [hunk 1]
```diff
@@ -1499,15 +1504,14 @@ public HFileBlock readBlock(long dataBlockOffset, long onDiskBlockSize,
...
           if (cacheBlock && cacheConf.shouldCacheBlockOnRead(category)) {
-            cachedRaw.set(cacheConf.shouldCacheCompressed(category));
-            cache.cacheBlock(cacheKey, cachedRaw.get() ? hfileBlock : unpacked,
+            cache.cacheBlock(cacheKey,
+              cacheConf.shouldCacheCompressed(category) ? hfileBlock : unpacked,
               cacheConf.isInMemory());
           }
         });
-        if (unpacked != hfileBlock && !cachedRaw.get()) {
+        if (unpacked != hfileBlock) {
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java]: Why this change?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java [hunk 2]
```diff
@@ -611,6 +609,10 @@ protected long evictBlock(LruCachedBlock block, boolean evictedByEvictionProcess
         victimHandler.cacheBlock(block.getCacheKey(), block.getBuffer());
       }
     }
+    // Decrease the block's reference count, and if refCount is 0, then it'll auto-deallocate. DO
+    // NOT move this up because if do that then the victimHandler may access the buffer with
+    // refCnt = 0 which is disallowed.
+    previous.getBuffer().release();
```
Comment (Apache9) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java]: So this is the problem. Mind explaining more? Why in victimHandler we will access the previous? And is it possible to add a UT?

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 44 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -0 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ HBASE-21879 Compile Tests _ |
| +1 | mvninstall | 278 | HBASE-21879 passed |
| +1 | compile | 55 | HBASE-21879 passed |
| +1 | checkstyle | 76 | HBASE-21879 passed |
| +1 | shadedjars | 280 | branch has no errors when building our shaded downstream artifacts. |
| -1 | findbugs | 168 | hbase-server in HBASE-21879 has 11 extant Findbugs warnings. |
| +1 | javadoc | 38 | HBASE-21879 passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 263 | the patch passed |
| +1 | compile | 54 | the patch passed |
| +1 | javac | 54 | the patch passed |
| -1 | checkstyle | 76 | hbase-server: The patch generated 1 new + 95 unchanged - 1 fixed = 96 total (was 96) |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 281 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 539 | Patch does not cause any errors with Hadoop 2.7.4 or 3.0.0. |
| +1 | findbugs | 188 | the patch passed |
| +1 | javadoc | 33 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 13822 | hbase-server in the patch failed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 17308 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.client.TestAsyncRegionAdminApi |
|   | hadoop.hbase.replication.TestReplicationDisableInactivePeer |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicas |
|   | hadoop.hbase.master.procedure.TestSCPWithReplicasWithoutZKCoordinated |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/242 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux fb443f933cf7 4.4.0-144-generic #170~14.04.1-Ubuntu SMP Mon Mar 18 15:02:05 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | HBASE-21879 / ab05d9d683 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| findbugs | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/artifact/out/branch-findbugs-hbase-server-warnings.html |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/testReport/ |
| Max. process+thread count | 5429 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-242/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java (hunk 1 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java]: Because no matter we put the compressedBlock or the uncompressedBlock into BlockCache, the BlockCache will retain the cached block (increasing its refCnt).  So here,  we only need to consider whether if the uncompressedBlock is excatly the compressedBlock, if not then just release the compressedBlock because the RPC will reference this uncompressedBlock. 
It's a bug which was introduced in HBASE-22005. so made a fix here.  
Will provide a UT for this.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java (hunk 2 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java]: > Why in victimHandler we will access the previous? And is it possible to add a UT?
For InclusiveCombinedBlockCache , we will move the evicted block from LRUCache to an larger  L2 cache (such as MemcachedBlockCache for longer caching I think),  So if release in line#596,  then victimHandler will cache an block which point to an unknown area because its memory has been free.
Yeah, will provide a UT.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java (hunk 1 above)
Comment (Apache9) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileReaderImpl.java]: Got it.

**Comment by openinx** [comment]
The patch still not fix the refCnt=0 retain issue, because I've applied this patch to my local branch , and deploy to my test cluster.  After run some days,  the QPS still dropped from 300000 Get/second to 200 Get/second.  Need to find out why.

**Comment by ramkrish86** [comment]
So even after this patch if QPS drops -  you stlil get the exception as attached in the original JIRA description?

**Comment by openinx** [comment]
bq. So even after this patch if QPS drops - you stlil get the exception as attached in the original JIRA description?
Yeah, still has the stacktrace. Let me check.
{code}
Mon May 20 12:16:36 CST 2019, RpcRetryingCaller{globalStartTime=1558325539029, pause=200, maxAttempts=16}, java.io.IOException: java.io.IOException: org.apache.hbase.thirdparty.io.netty.util.IllegalReferenceCountException: refCnt: 0, increment: 1
	at org.apache.hadoop.hbase.regionserver.HRegion$RegionScannerImpl.handleException(HRegion.java:6534)
	at org.apache.hadoop.hbase.regionserver.HRegion$RegionScannerImpl.initializeScanners(HRegion.java:6504)
	at org.apache.hadoop.hbase.regionserver.HRegion$RegionScannerImpl.<init>(HRegion.java:6473)
	at org.apache.hadoop.hbase.regionserver.HRegion.instantiateRegionScanner(HRegion.java:2999)
	at org.apache.hadoop.hbase.regionserver.HRegion.getScanner(HRegion.java:2979)
	at org.apache.hadoop.hbase.regionserver.HRegion.getScanner(HRegion.java:2961)
	at org.apache.hadoop.hbase.regionserver.HRegion.getScanner(HRegion.java:2955)
	at org.apache.hadoop.hbase.regionserver.RSRpcServices.get(RSRpcServices.java:2621)
	at org.apache.hadoop.hbase.regionserver.RSRpcServices.get(RSRpcServices.java:2548)
	at org.apache.hadoop.hbase.shaded.protobuf.generated.ClientProtos$ClientService$2.callBlockingMethod(ClientProtos.java:41998)
	at org.apache.hadoop.hbase.ipc.RpcServer.call(RpcServer.java:374)
	at org.apache.hadoop.hbase.ipc.CallRunner.run(CallRunner.java:132)
	at org.apache.hadoop.hbase.ipc.RpcExecutor$Handler.run(RpcExecutor.java:324)
	at org.apache.hadoop.hbase.ipc.RpcExecutor$Handler.run(RpcExecutor.java:304)
Caused by: org.apache.hbase.thirdparty.io.netty.util.IllegalReferenceCountException: refCnt: 0, increment: 1
	at org.apache.hbase.thirdparty.io.netty.util.AbstractReferenceCounted.retain0(AbstractReferenceCounted.java:87)
	at org.apache.hbase.thirdparty.io.netty.util.AbstractReferenceCounted.retain(AbstractReferenceCounted.java:74)
	at org.apache.hadoop.hbase.nio.SingleByteBuff.retain(SingleByteBuff.java:398)
	at org.apache.hadoop.hbase.nio.SingleByteBuff.retain(SingleByteBuff.java:39)
	at org.apache.hadoop.hbase.io.hfile.HFileBlock.retain(HFileBlock.java:449)
	at org.apache.hadoop.hbase.io.hfile.HFileBlock.retain(HFileBlock.java:115)
	at org.apache.hadoop.hbase.io.hfile.LruBlockCache.getBlock(LruBlockCache.java:538)
	at org.apache.hadoop.hbase.io.hfile.CombinedBlockCache.getBlock(CombinedBlockCache.java:84)
	at org.apache.hadoop.hbase.io.hfile.HFileReaderImpl.getCachedBlock(HFileReaderImpl.java:1298)
	at org.apache.hadoop.hbase.io.hfile.HFileReaderImpl.readBlock(HFileReaderImpl.java:1464)
	at org.apache.hadoop.hbase.io.hfile.HFileBlockIndex$CellBasedKeyBlockIndexReader.loadDataBlockWithScanInfo(HFileBlockIndex.java:339)
	at org.apache.hadoop.hbase.io.hfile.HFileReaderImpl$HFileScannerImpl.seekTo(HFileReaderImpl.java:844)
	at org.apache.hadoop.hbase.io.hfile.HFileReaderImpl$HFileScannerImpl.seekTo(HFileReaderImpl.java:794)
	at org.apache.hadoop.hbase.regionserver.StoreFileScanner.seekAtOrAfter(StoreFileScanner.java:315)
	at org.apache.hadoop.hbase.regionserver.StoreFileScanner.seek(StoreFileScanner.java:216)
	at org.apache.hadoop.hbase.regionserver.StoreScanner.seekScanners(StoreScanner.java:394)
	at org.apache.hadoop.hbase.regionserver.StoreScanner.<init>(StoreScanner.java:249)
	at org.apache.hadoop.hbase.regionserver.HStore.createScanner(HStore.java:2063)
	at org.apache.hadoop.hbase.regionserver.HStore.getScanner(HStore.java:2054)
	at org.apache.hadoop.hbase.regionserver.HRegion$RegionScannerImpl.initializeScanners(HRegion.java:6493)
	... 12 more
{code}

**Comment by openinx** [comment]
Update the patch with the final fix and UT, FYI @Apache9 @ramkrish86 @anoopsjohn

**Comment by openinx** [comment]
Seems no hadoop QA feedback ?  It's strange...

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlockIndex.java [hunk 3]
```diff
@@ -313,10 +313,13 @@ public BlockWithScanInfo loadDataBlockWithScanInfo(Cell key, HFileBlock currentB
...
 
       HFileBlock block = null;
-      boolean dataBlock = false;
       KeyOnlyKeyValue tmpNextIndexKV = new KeyValue.KeyOnlyKeyValue();
       while (true) {
         try {
+          // Must initialize it with null here, because if don't and once an exception happen in
+          // readBlock, then we'll release the previous assigned block twice in the finally block.
+          // (See HBASE-22422)
+          block = null;
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlockIndex.java]: Good catch.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/BlockCacheUtil.java [hunk 4]
```diff
@@ -230,6 +230,9 @@ public static boolean shouldReplaceExistingCacheBlock(BlockCache blockCache,
       BlockCacheKey cacheKey, Cacheable newBlock) {
     // NOTICE: The getBlock has retained the existingBlock inside.
     Cacheable existingBlock = blockCache.getBlock(cacheKey, false, false, false);
+    if (existingBlock == null) {
```
[the remark this answers is not in the document]
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/BlockCacheUtil.java]: Yes, see the LruBlockCache#cacheBlock: 
```
    LruCachedBlock cb = map.get(cacheKey);
    if (cb != null && !BlockCacheUtil.shouldReplaceExistingCacheBlock(this, cacheKey, buf)) {
      return;
    }
```
The existence pre-check and accessing block in shouldReplaceExistingCacheBlock is not atomic op.  so if any eviction happen between them,  the NPE will happen.  I added a  UT testMultiThreadGetAndEvictBlock to address this.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java [hunk 5]
```diff
@@ -1533,21 +1534,28 @@ public boolean containsKey(BlockCacheKey key) {
...
         entry.getData().retain();
-      }
-      return previous;
+        absent.set(true);
+        return entry;
+      });
+      return absent.get() ? null : re;
     }
 
     public boolean remove(BlockCacheKey key) {
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java]: While removing the atomicity is not needed? Because we do computeIfAbsent and that is already now atomically guarded?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java [hunk 6]
```diff
@@ -1533,21 +1534,28 @@ public boolean containsKey(BlockCacheKey key) {
...
+      AtomicBoolean absent = new AtomicBoolean(false);
+      RAMQueueEntry re = delegate.computeIfAbsent(key, k -> {
         // The RAMCache reference to this entry, so reference count should be increment.
         entry.getData().retain();
-      }
-      return previous;
+        absent.set(true);
+        return entry;
+      });
+      return absent.get() ? null : re;
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java]: I think I got it now. You have changed the get() to computeIfPresent(). So if the remove has already removed the entry then the get() cannot do a retain(). So a similar change is also needed for putIfAbsent() also? Rest looks good to me.

Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java]: Yeah,  if don't make the put and retain in atomic , then if remove & release happen between the put and retain, finally we 're retain a block with refCnt=0 which is also disallowed.  Thanks.
